lib/rex/post/meterpreter/packet.rb

```diff
@@ -693,17 +819,31 @@ def to_r
   # passing it on to the default functionality that can parse
   # the TLV values.
   #
-  def from_r(bytes)
-    xor_key = bytes[0,4].unpack('N')[0]
-    super(xor_bytes(xor_key, bytes[4, bytes.length]))
+  def from_r(aes_key=nil)
+    xor_key = self.raw[XOR_KEY_OFF, XOR_KEY_SIZE]
+    data = xor_key + xor_bytes(xor_key, self.raw[SESSION_GUID_OFF, self.raw.length - SESSION_GUID_OFF])
+    self.session_guid = data[SESSION_GUID_OFF, SESSION_GUID_SIZE]
+    encrypted_flag = data[ENCRYPTED_FLAG_OFF, ENCRYPTED_FLAG_SIZE] == "\x01"
```

Comment:
Good place to mention - is it worth designating the `ENCRYPTED_FLAG` field as `FLAGS` with the bottom bit being `ENCRYPTED_FLAG` and the other 7 bits being reserved? Will 7 be enough for future use? In which case we should use boolean logic to extract the bottom bit rather than just `== "\x01"` to pave the way for general flag field use